backend/src/apiserver/template/template_test.go

```diff
@@ -300,3 +246,63 @@ func TestModelToCRDTrigger_Cron(t *testing.T) {
 	assert.Nil(t, err)
 	assert.Equal(t, expectedCRDTrigger, actualCRDTrigger)
 }
+
+func loadYaml(t *testing.T, path string) string {
+	res, err := ioutil.ReadFile(path)
+	if err != nil {
+		t.Error(err)
+	}
+	return string(res)
+}
+
+func TestIsKubernetesExecutorConfig(t *testing.T) {
+	template := loadYaml(t, "testdata/pipeline_with_volume.yaml")
+	splitTemplate := strings.Split(template, "\n---\n")
+	assert.True(t, isKubernetesExecutorConfig([]byte(splitTemplate[1])))
+	assert.False(t, isKubernetesExecutorConfig([]byte(splitTemplate[0])))
+}
+
+func TestNewTemplate_V2(t *testing.T) {
+	template := loadYaml(t, "testdata/hello_world.yaml")
+	expectedSpecJson, _ := yaml.YAMLToJSON([]byte(template))
+	var expectedSpec pipelinespec.PipelineSpec
+	protojson.Unmarshal(expectedSpecJson, &expectedSpec)
+	expectedTemplate := &V2Spec{
+		spec: &expectedSpec,
+	}
+	templateV2Spec, err := New([]byte(template))
+	assert.Nil(t, err)
+	assert.Equal(t, expectedTemplate, templateV2Spec)
+}
+
+func TestNewTemplate_WithExecutorConfig(t *testing.T) {
+	template := loadYaml(t, "testdata/pipeline_with_volume.yaml")
+	var expectedPipelineSpec pipelinespec.PipelineSpec
+	var expectedPlatformSpec pipelinespec.PlatformSpec
+
+	splitTemplate := strings.Split(template, "\n---\n")
+	expectedSpecJson, _ := yaml.YAMLToJSON([]byte(splitTemplate[0]))
+	protojson.Unmarshal(expectedSpecJson, &expectedPipelineSpec)
+
+	expectedPlatformSpecJson, _ := yaml.YAMLToJSON([]byte(splitTemplate[1]))
+	protojson.Unmarshal(expectedPlatformSpecJson, &expectedPlatformSpec)
+
+	expectedTemplate := &V2Spec{
+		spec:         &expectedPipelineSpec,
+		platformSpec: &expectedPlatformSpec,
+	}
+	templateV2Spec, err := New([]byte(template))
+	assert.Nil(t, err)
+	assert.Equal(t, expectedTemplate, templateV2Spec)
+}
+
+// Verify that the V2Spec object created from Bytes() method is the same as the original object.
+// The byte slice may be slightly different from the original input during the conversion.
+func TestBytes_V2(t *testing.T) {
+	template := loadYaml(t, "testdata/pipeline_with_volume.yaml")
+	templateV2Spec, _ := New([]byte(template))
+	templateBytes := templateV2Spec.Bytes()
+	newTemplateV2Spec, err := New(templateBytes)
+	assert.Nil(t, err)
+	assert.Equal(t, templateV2Spec, newTemplateV2Spec)
```

Comment:
If I got it right, the purpose of `Bytes()` method is to return the []byte of the original spec YAML?

Then, we need to compare `assert.Equal(t, template, string(templateBytes))`

Reply:
I understand that it would be best to have `Bytes()` return the exact same content as the spec YAML, but with the marshaling and unmarshaling, it is quite inevitable that there will be some small changes, such as single quote being replaced by double quote. That's why I resorted to comparing the unmarshaled struct. I also think it's good practice to compare the struct due to the flakiness of the serialized format.